internal/cloud/gcp/cloud.go

```diff
@@ -0,0 +1,319 @@
+	var closers []func() error
+
+	insAPI, err := compute.NewInstancesRESTClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	closers = append(closers, insAPI.Close)
+	forwardingRulesAPI, err := compute.NewGlobalForwardingRulesRESTClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	closers = append(closers, forwardingRulesAPI.Close)
+	subnetAPI, err := compute.NewSubnetworksRESTClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	closers = append(closers, subnetAPI.Close)
+
+	return &Cloud{
+		imds:               imds.NewClient(nil),
+		instanceAPI:        &instanceClient{insAPI},
+		forwardingRulesAPI: &forwardingRulesClient{forwardingRulesAPI},
+		subnetAPI:          subnetAPI,
+		closers:            closers,
+	}, nil
+}
+
+// Close closes all connections to the GCP API server.
+func (c *Cloud) Close() {
+	for _, close := range c.closers {
+		_ = close()
+	}
+}
+
+// GetInstance retrieves an instance using its providerID.
+func (c *Cloud) GetInstance(ctx context.Context, providerID string) (metadata.InstanceMetadata, error) {
+	project, zone, instanceName, err := gcpshared.SplitProviderID(providerID)
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("invalid providerID: %w", err)
+	}
+
+	return c.getInstance(ctx, project, zone, instanceName)
+}
+
+// GetLoadBalancerEndpoint returns the endpoint of the load balancer.
+func (c *Cloud) GetLoadBalancerEndpoint(ctx context.Context) (string, error) {
+	project, zone, instanceName, err := c.retrieveInstanceInfo()
+	if err != nil {
+		return "", err
+	}
+	uid, err := c.uid(ctx, project, zone, instanceName)
+	if err != nil {
+		return "", err
+	}
+
+	var resp *computepb.ForwardingRule
+	iter := c.forwardingRulesAPI.List(ctx, &computepb.ListGlobalForwardingRulesRequest{
+		Project: project,
+		Filter:  proto.String(fmt.Sprintf("(labels.%s:%s) AND (labels.%s:kubernetes)", cloud.TagUID, uid, tagUsage)),
+	})
+	for resp, err = iter.Next(); err == nil; resp, err = iter.Next() {
+		if resp.PortRange == nil {
+			continue
+		}
+		if resp.IPAddress == nil {
+			continue
+		}
+		portRange := strings.Split(*resp.PortRange, "-")
+		return net.JoinHostPort(*resp.IPAddress, portRange[0]), nil
+	}
+
+	return "", fmt.Errorf("kubernetes load balancer with UID %s not found: %w", uid, err)
+}
+
+// List retrieves all instances belonging to the current constellation.
+func (c *Cloud) List(ctx context.Context) ([]metadata.InstanceMetadata, error) {
+	project, zone, instanceName, err := c.retrieveInstanceInfo()
+	if err != nil {
+		return nil, err
+	}
+	uid, err := c.uid(ctx, project, zone, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	var instances []metadata.InstanceMetadata
+	var resp *computepb.Instance
+	iter := c.instanceAPI.List(ctx, &computepb.ListInstancesRequest{
+		Filter:  proto.String(fmt.Sprintf("labels.%s:%s", cloud.TagUID, uid)),
+		Project: project,
+		Zone:    zone,
+	})
+	for resp, err = iter.Next(); err == nil; resp, err = iter.Next() {
```

Comment:
Imo, the opposite is true.
Moving break conditions and iterator operations into an infinite loop is unnecessary if the language already provides a proper way to handle things.

`resp` has to be declared outside the loop because we want to access `err` outside of the loop. Otherwise the `for resp, err := iter.Next();` would scope the error to only the loop.

Reply:
I'm also fine with the current way. 

I know, that's why I designed my suggestion in a way that doesn't need `err` outside of the for loop.